ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs

```diff
@@ -587,6 +588,32 @@ releaseResources rr keys releaser = do
 withRegistry :: (IOLike m, HasCallStack) => (ResourceRegistry m -> m a) -> m a
 withRegistry = bracket unsafeNewRegistry closeRegistry
 
+-- | Create a new registry as part of a standard 'bracket'
+--
+-- Use this combinator to emphasize that the lifetime of the registry exactly
+-- spans the lifetime of the explicitly bracketed resources, no more and no
+-- less.
+--
+-- The only difference from 'bracketWithRegistry_' is that the explicitly
+-- bracketed resources do rely on the registry.
+--
+-- See documentation of 'ResourceRegistry' for a more general discussion.
+bracketWithRegistry ::
+     (IOLike m, HasCallStack)
+  => (ResourceRegistry m -> m a)
+     -- ^ Allocate the bracketed resources. They depend on the registry but are
+     -- not themselves registered, since they're simply 'bracket'ed.
+  -> (a -> m ())
+     -- ^ Release the bracketed resources.
+  -> (ResourceRegistry m -> a -> m r)
+  -> m r
+bracketWithRegistry newA closeA body =
+    -- It would be OK to fuse these two calls to 'bracket', but it's simpler and
+    -- less error-prone this way.
+    withRegistry $ \registry ->
+      bracket (newA registry) closeA $ \a ->
```

Comment:
I noticed this pr and #2892 while working on some changes at the ChainSync that we need for db-sync. I think in general blending brackets and registries can be tricky, basically because of the order of deallocation. Here `a` will be freed before anything else allocated with the registry. By using a simple registry instead, things are freed from younger to older.

Reply:
I'm not sure if this is a real issue for Reader/Follower, but may be worth at least documenting it for `bracketWithRegistry`.

Reply:
The point is that this is a _new_ registry: we are constructing something that for some reason needs a registry (example: the reader), but really the scope of that thing and the registry are coupled. Notice that _nothing else_ has access to this registry: the only thing that has access to it is the new resource itself. Hence, deallocating that resource before anything else in the registry is okay. 

What you are saying is true nonetheless: _if_ we then later allocate a new resource that somehow refers to this new resource (i.e., a reference in the reverse direction), we might be in trouble. That kind of "cyclic dependency" graph would indeed be potentially problematic.

Reply:
Yes this is the usual case: newer resources reference older resources. That's why the younger to older deallocation that's done by Registry is the normal thing to do. And it's a kinda violated by the bracket.

Reply:
Yes, you are correct. It would indeed be good to state the side condition here: those newer conditions should not be referring to the resources that is allocated by this new abstraction.

Reply:
The point is that the lifetime of this new resource _is equal to_ the lifetime of the registry.

Reply:
Good observation!

Reply:
> The point is that the lifetime of this new resource is equal to the lifetime of the registry.

For future reference, there are some minor caveats to that; see the comment on `bracketWithPrivateRegistry`.